zcash_primitives/src/sapling/prover.rs

```diff
@@ -1,16 +1,75 @@
 //! Abstractions over the proving system and parameters.
 
+use rand_core::RngCore;
+
 use crate::{
     sapling::{
         self,
         redjubjub::{PublicKey, Signature},
-        value::ValueCommitment,
+        value::{NoteValue, ValueCommitTrapdoor, ValueCommitment},
+        MerklePath,
     },
     transaction::components::{Amount, GROTH_PROOF_SIZE},
 };
 
 use super::{Diversifier, PaymentAddress, ProofGenerationKey, Rseed};
 
+/// Interface for creating Sapling Spend proofs.
+pub trait SpendProver {
+    /// The proof type created by this prover.
+    type Proof;
```

Comment:
Technically now that `zcash_primitives` depends on `bellman` (after rebasing in #999), we could remove this associated type and just use `Proof<Bls12>`, but to some extent that does depend on the trait implementation (e.g. a remote prover might just return bytes here).